tests/next/include/lib.sh

```diff
@@ -0,0 +1,128 @@
+#!/bin/bash
+
+# Allow more verbosity for debugging purposes.
+[ "$V" == "1" ] && set -x
+
+CLEANUP_LIST=()
+
+# Add a command to be run at cleanup time.
+cleanup() {
+	CLEANUP_LIST+=("$@")
+}
+
+# Cleanup all resources. Called between each test.
+cleanup_all() {
+	for ((i=${#CLEANUP_LIST[@]} - 1; i >= 0; i--)); do
+		${CLEANUP_LIST[$i]}
+	done
+	CLEANUP_LIST=()
+}
+
+# Run the Retis debug binary.
+retis=$(git rev-parse --show-toplevel)/target/debug/retis
+[ ! -x $retis ] && { echo "Please build Retis first ('make')"; exit 1; }
+
+# Allow importing custom Python modules.
+export PYTHONPATH="${PYTHONPATH}:$(git rev-parse --show-toplevel)/tests/next/include"
+
+# Create network namespaces. Take a list of network namespace names as the
+# input.
+add_ns() {
+	for ns in $@; do
+		ip netns add $ns
+		ip -n $ns link set lo up
+		cleanup "ip netns del $ns | xargs -r kill"
+	done
+}
+
+# Add a veth pair.
+add_veth_pair() {
+	veth0_name=$1
+	veth0_ns=$2	# Optional ("" to skip)
+	veth1_name=$3
+	veth1_ns=$4	# Optional ("" to skip)
+
+	[ ! -z "$veth0_ns" ] && ns0="netns $veth0_ns" || true
+	[ ! -z "$veth1_ns" ] && ns1="netns $veth1_ns" || true
+
+	ip link add $veth0_name $ns0 type veth peer name $veth1_name $ns1
+	ip -net ns0 link set $veth0_name up
+	ip -net ns1 link set $veth1_name up
+
+	# Cleanup interfaces in the main netns.
+	[ -z "$veth0_ns" ] && cleanup "ip link del $veth0_name" || true
+	[ -z "$veth1_ns" ] && cleanup "ip link del $veth1_name" || true
+}
+
+# Allow capturing stdout/stderr while running tests. Tests should exit on
+# errors so we can't redirect the test function outputs directly and have to
+# redirect the main stdout/stderr fds. We also have to keep additional fds
+# pointing to the original stdout/stderr fds for restoring things after a test
+# completes or fails.
+exec 21>&1; exec 22>&2
+__capture_stdout() {
+	[ "$V" != "1" ] && exec 1>log 2>&1 || true
+}
+__restore_stdout() {
+	exec 1>&21 2>&22
+}
+
+# Run a list of tests. Takes the test functions as an argument.
+run_tests() {
+	# Allow overriding the tests list to run.
+	tests="${TESTS:-$@}"
+	for t in $tests; do
+		# Do not error on unknown tests to allow `TESTS` use with
+		# `make functional-tests`.
+		typeset -f $t >/dev/null || continue
+		[ "$LIST_TESTS" == "1" ] && { echo $t; continue; }
+
+		tmpdir=$(mktemp -d -t retis-test-XXXX)
+		cleanup "rm -rf $tmpdir"
+		cd $tmpdir
+
+		echo -n "Running $t test..."
```

Comment:
Have you considered printing stuff in a standard-ish format? e.g: tap or ktap?

Reply:
I think tap is a fairly easy and flexible format. I was planning to output the result of the current functional tests in junit for cirrus to consume them directly and show each individual test pass/fail in the GH interface. Doing junit in bash might be an overkill but there might be tap-2-junit converters (have not looked).